### DISCUSSION TITLE
Issue with Redundant Nodes After Merging Projects

### DISCUSSION THREAD

**Discussion by heiley8181** [talmolab/sleap#1700] [Help!] [answered]
Hello SLEAP team,

I am seeking guidance to resolve the issue I am currently facing right now. I recently combined two projects and have encountered an issue with redundant nodes. 

![image](https://github.com/talmolab/sleap/assets/76165836/7c849303-ddb9-49e3-a987-2c566c50399d)

As you can see, Tail2, Body_R and Body_L are the nodes that I no longer use. I reviewed the original project files to eliminate these nodes, yet they constantly reappear after merging. 

![image](https://github.com/talmolab/sleap/assets/76165836/b59f6fa4-4d33-4b2d-8eed-2e9148a11716)
The problem is that these nodes are not visible in the skeleton window as well.

It would be super helpful if anyone could tell me what I should try. It seems like very simple issue but I just cannot figure out why :(

Thank you so much in advance!

**Comment by heiley8181** [comment]
Hi SLEAP team,

So I found this super helpful discussion https://github.com/talmolab/sleap/issues/713#issue-1217102006 and checked the structure of my nodes for each project that I am trying to merge.

![image](https://github.com/talmolab/sleap/assets/76165836/cd6e0801-3439-4cff-ae39-5652a2dcc242)

So it seems like I have unwanted nodes in both files (which are Body_L, Body_R, and Tail2) but for some reason, I cannot delete them on the SLEAP gui. Is there any way to delete them using iPython? The redundant nodes do not show up on the skeleton window for both files.
Let me know what I can try next.

Thank you :)

**Reply by Marielamend** [reply]
Hello @heiley8181!
So sorry for the delayed response, 
Some follow-up questions:
- Which version of SLEAP are you using?
- Are you receiving any error messages?

Have a great day, 
Mariela

**Reply by heiley8181** [reply]
Hi @Marielamend,
Thanks for the response :)
1. I am using 1.3.3 version
2. When I open the inferenceonly5nodes.slp, I don't get any error messages but when I open the second file, this is what I get: 

```
Restoring GUI state...
Traceback (most recent call last):
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 287, in openProject
    self.execute(OpenProject, filename=filename, first_open=first_open)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 241, in execute
    command().execute(context=self, params=kwargs)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 137, in execute
    self.do_with_signal(context, params)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 161, in do_with_signal
    cls.do_action(context, params)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 723, in do_action
    new_window.commands.loadProjectFile(filename)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 272, in loadProjectFile
    self.execute(LoadProjectFile, filename=filename)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 241, in execute
    command().execute(context=self, params=kwargs)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 137, in execute
    self.do_with_signal(context, params)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 161, in do_with_signal
    cls.do_action(context, params)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 674, in do_action
    context.app.on_data_update([UpdateTopic.project, UpdateTopic.all])
  File "c:\windows\system32\sleap\sleap\gui\app.py", line 1176, in on_data_update
    self.skeleton_dock.nodes_table.model().items = self.state["skeleton"]
  File "c:\windows\system32\sleap\sleap\gui\dataviews.py", line 103, in items
    item_data = self.item_to_data(obj, item)
  File "c:\windows\system32\sleap\sleap\gui\dataviews.py", line 405, in item_to_data
    return dict(name=item.name, symmetry=obj.get_symmetry_name(item.name))
  File "c:\windows\system32\sleap\sleap\skeleton.py", line 798, in get_symmetry_name
    symmetric_node = self.get_symmetry(node)
  File "c:\windows\system32\sleap\sleap\skeleton.py", line 774, in get_symmetry
    node_node = self.find_node(node)
  File "c:\windows\system32\sleap\sleap\skeleton.py", line 573, in find_node
    raise ValueError("Found multiple nodes named ({}).".format(name))
ValueError: Found multiple nodes named (Body_R).
```

Could you let me know what I can try to get rid of the multiple nodes?

Thank you so much!

-Heiley

**Reply by Marielamend** [reply]
Hello @heiley8181!

Thank you for your response, it seems like this is a bug. Let me label it as such and if you can, [upload both of your slp files in this form](https://forms.gle/KebBPdjoHw2WMKmV8) so we can best tackle this!

Best, 
Mariela

**Reply by heiley8181** [reply]
Hello, @Marielamend 

I just wanted to check if there were any updates regarding this issue. I noticed that whenever I try to merge the projects, instead of getting `ValueError: Found multiple nodes named (Body_R)`, now I get a different error message: 

```
Restoring GUI state...
Traceback (most recent call last):
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 604, in mergeProject
    self.execute(MergeProject, filenames=filenames)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 241, in execute
    command().execute(context=self, params=kwargs)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 133, in execute
    self.ask_and_do(context, params)
  File "c:\windows\system32\sleap\sleap\gui\commands.py", line 2834, in ask_and_do
    MergeDialog(base_labels=context.labels, new_labels=new_labels).exec_()
  File "c:\windows\system32\sleap\sleap\gui\dialogs\merge.py", line 51, in __init__
    if self.base_labels.skeleton.node_names != self.new_labels.skeleton.node_names:
  File "c:\windows\system32\sleap\sleap\io\dataset.py", line 564, in skeleton
    "Labels.skeleton can only be used when there is only a single skeleton "
ValueError: Labels.skeleton can only be used when there is only a single skeleton saved in the labels. Use Labels.skeletons instead.
```

If possible, could you recommend what I should try to successfully merge the projects without having any redundant nodes? Thank you so much in advance!

**Reply by eberrigan** [reply]
Hi @heiley8181. I have your files and I am working on this and will try to get an answer to you by the end of the day! Thanks for your patience.

**Reply by eberrigan** [reply]
Hi @heiley8181 I do not have your videos so I am not able to open this in the GUI. Are you able to upload those as well? 

From the side of the API, it appears that when after you delete the nodes they are still in the skeleton. I am able to set the skeletons equal to each other and then merge and save a new file with the desired skeleton. Please see this code below. 

If the GUI skeleton window is not accurately reflecting skeleton changes this is an issue we should address. Here is the work around for now. Please let me know if you have any questions. 

```
import sleap
import pandas as pd
import numpy as np

from pathlib import Path

# Define labels paths
labels_5nodes = Path("inferenceonly5nodes.slp")
labels_notail = Path("mergedwhiteonlynotail.slp")

# Load the labels
labels_5nodes = sleap.load_file(str(labels_5nodes))
labels_notail = sleap.load_file(str(labels_notail))

print(f"Labels with 5 nodes: {labels_5nodes}")
print(f"Labels with no tail: {labels_notail}")

# Get the skeletons
labels_5nodes_skeletons = labels_5nodes.skeletons
labels_notail_skeletons = labels_notail.skeletons

print(f"Skeletons with 5 nodes: {labels_5nodes_skeletons}")
print(f"Skeletons with no tail: {labels_notail_skeletons}")

print(f"Skeletons with 5 nodes is the same as skeletons with no tail: {labels_5nodes_skeletons==labels_notail_skeletons}")

# This sets skeletons equal to another
labels_notail_skeletons = labels_5nodes_skeletons

print(f"Skeletons with 5 nodes: {labels_5nodes_skeletons}")
print(f"Skeletons with no tail after modification: {labels_notail_skeletons}")
print(f"Now the skeletons should be the same: {labels_5nodes_skeletons==labels_notail_skeletons}")

# So now we can merge them since they have the same skeletons
labels_5nodes.extend_from(labels_notail, unify=True)
print(f"Merged labels: {labels_5nodes}")

# Save the merged labels
labels_5nodes.save("merged_labels.slp")
print(f"saved merged_labels.slp")
```

**Reply by heiley8181** [reply]
Hi @eberrigan ,

Thank you so much for the assistance. However, I still see the unwanted nodes in the new "merged_labels.slp" file, and they aren't visible in the skeleton window. 

I have a total of 19 video files, but unfortunately, the compressed file isn't uploading to the Google form provided above. Please let me know if there are alternative links where I can share my video files.

Thank you very much for your continued help :)

**Reply by eberrigan** [reply]
Hello,

Sorry to hear that. 

Yes you can use Google drive or something similar and send me the link at eberrigan@salk.edu.

Thanks!

Elizabeth

**Reply by eberrigan** [reply]
Hi @heiley8181, 

Now that I have your video files, I noticed that I have a labels file and the merged labels. Can you also provide the labels you merged with since this should be the file with the different skeleton?

Here is a notebook that may help you resolve the problem as well. Since we have modified the skeleton, we need to update the instances to remove the extra nodes. https://colab.research.google.com/drive/1OmnmZIM64pWFqoeNkUm2QnpOWmPwoUbN#scrollTo=efYogvQi0ta3 

Thanks!

Elizabeth

**Reply by heiley8181** [reply]
Hi @eberrigan ,

I am not sure if I understood you correctly, but both files should have same skeletons before merging. The problem is that when I select 'merge into project' within the mergedwhiteonlynotail.slp and use inferenceonly5nodes.slp, discard all conflicting instances, I get the unwanted nodes reappearing on the merged files. The skeletons shouldn't be different before merging them.

**Reply by eberrigan** [reply]
Thanks for explaining that @heiley8181. So did you merge files to make mergedwhiteonlynotail.slp? Or did you just delete nodes after labeling?

**Reply by heiley8181** [reply]
Hi @eberrigan ,

I think I merged some files to make 'mergedwhiteonlynotail.slp' but I am not sure which one I used. I also deleted the unwanted nodes like Body_L, Body_R, and Tail2 from the skeleton GUI to make the mergedwhiteonlynotail.slp file.

**Reply by eberrigan** [reply, marked as answer]
This should be fixed by #2075 when it is merged.
